In [1]:
import sys
from pathlib import Path

import pandas as pd
import numpy as np

PROJECT_ROOT = Path.cwd().parent
sys.path.append(str(PROJECT_ROOT))

from src.data.data_loader import load_application_train

df = load_application_train()

print("Dataset shape:", df.shape)
print("Columns:", len(df.columns))

Dataset shape: (307511, 122)
Columns: 122


In [2]:
df["LOAN_INCOME_RATIO"] = (
    df["AMT_CREDIT"] / df["AMT_INCOME_TOTAL"]
)

df["ANNUITY_INCOME_RATIO"] = (
    df["AMT_ANNUITY"] / df["AMT_INCOME_TOTAL"]
)

df["CREDIT_GOODS_RATIO"] = (
    df["AMT_CREDIT"] / df["AMT_GOODS_PRICE"]
)

print(
    df[
        [
            "LOAN_INCOME_RATIO",
            "ANNUITY_INCOME_RATIO",
            "CREDIT_GOODS_RATIO"
        ]
    ].describe()
)

       LOAN_INCOME_RATIO  ANNUITY_INCOME_RATIO  CREDIT_GOODS_RATIO
count      307511.000000         307499.000000       307233.000000
mean            3.957570              0.180930            1.122995
std             2.689728              0.094574            0.124045
min             0.004808              0.000224            0.150000
25%             2.018667              0.114782            1.000000
50%             3.265067              0.162833            1.118800
75%             5.159880              0.229067            1.198000
max            84.736842              1.875965            6.000000


C:\Users\Dell\AppData\Local\Temp\ipykernel_19220\3042369113.py:1: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df["LOAN_INCOME_RATIO"] = (
C:\Users\Dell\AppData\Local\Temp\ipykernel_19220\3042369113.py:5: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df["ANNUITY_INCOME_RATIO"] = (
C:\Users\Dell\AppData\Local\Temp\ipykernel_19220\3042369113.py:9: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once u

In [3]:
df["AGE_YEARS"] = (
    -df["DAYS_BIRTH"] / 365.25
)

df["DAYS_EMPLOYED_CLEAN"] = (
    df["DAYS_EMPLOYED"].replace(365243, np.nan)
)

df["EMPLOYMENT_YEARS"] = (
    -df["DAYS_EMPLOYED_CLEAN"] / 365.25
)

print(
    df[
        [
            "AGE_YEARS",
            "EMPLOYMENT_YEARS"
        ]
    ].describe()
)

           AGE_YEARS  EMPLOYMENT_YEARS
count  307511.000000     252137.000000
mean       43.906900          6.527500
std        11.947950          6.402081
min        20.503765         -0.000000
25%        33.984942          2.099932
50%        43.121150          4.511978
75%        53.886379          8.692676
max        69.073238         49.040383


C:\Users\Dell\AppData\Local\Temp\ipykernel_19220\150445333.py:1: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df["AGE_YEARS"] = (
C:\Users\Dell\AppData\Local\Temp\ipykernel_19220\150445333.py:5: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df["DAYS_EMPLOYED_CLEAN"] = (
C:\Users\Dell\AppData\Local\Temp\ipykernel_19220\150445333.py:9: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.conc

In [4]:
df["INCOME_PER_CHILD"] = (
    df["AMT_INCOME_TOTAL"] / (df["CNT_CHILDREN"] + 1)
)

df["INCOME_PER_FAMILY_MEMBER"] = (
    df["AMT_INCOME_TOTAL"] / (df["CNT_FAM_MEMBERS"] + 1)
)

print(
    df[
        [
            "INCOME_PER_CHILD",
            "INCOME_PER_FAMILY_MEMBER"
        ]
    ].describe()
)

       INCOME_PER_CHILD  INCOME_PER_FAMILY_MEMBER
count      3.075110e+05              3.075090e+05
mean       1.395079e+05              5.748225e+04
std        1.458110e+05              6.585166e+04
min        3.000000e+03              2.647059e+03
25%        7.875000e+04              3.375000e+04
50%        1.170000e+05              4.950000e+04
75%        1.800000e+05              6.750000e+04
max        5.850000e+07              2.925000e+07


C:\Users\Dell\AppData\Local\Temp\ipykernel_19220\2002747403.py:1: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df["INCOME_PER_CHILD"] = (
C:\Users\Dell\AppData\Local\Temp\ipykernel_19220\2002747403.py:5: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df["INCOME_PER_FAMILY_MEMBER"] = (


In [5]:
df["CREDIT_MINUS_INCOME"] = (
    df["AMT_CREDIT"] - df["AMT_INCOME_TOTAL"]
)

df["ANNUITY_MINUS_INCOME"] = (
    df["AMT_ANNUITY"] - df["AMT_INCOME_TOTAL"]
)

print(
    df[
        [
            "CREDIT_MINUS_INCOME",
            "ANNUITY_MINUS_INCOME"
        ]
    ].describe()
)

       CREDIT_MINUS_INCOME  ANNUITY_MINUS_INCOME
count         3.075110e+05          3.074990e+05
mean          4.302281e+05         -1.416887e+05
std           4.339157e+05          2.347809e+05
min          -1.164375e+08         -1.169738e+08
25%           1.422000e+05         -1.732500e+05
50%           3.461760e+05         -1.210770e+05
75%           6.340410e+05         -8.544150e+04
max           3.645000e+06          2.567700e+04


C:\Users\Dell\AppData\Local\Temp\ipykernel_19220\701523266.py:1: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df["CREDIT_MINUS_INCOME"] = (
C:\Users\Dell\AppData\Local\Temp\ipykernel_19220\701523266.py:5: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df["ANNUITY_MINUS_INCOME"] = (


In [6]:
df["EMPLOYMENT_AGE_RATIO"] = (
    df["EMPLOYMENT_YEARS"] / df["AGE_YEARS"]
)

print(
    df["EMPLOYMENT_AGE_RATIO"].describe()
)

count    252137.000000
mean          0.156861
std           0.133549
min          -0.000000
25%           0.056099
50%           0.118733
75%           0.219170
max           0.728811
Name: EMPLOYMENT_AGE_RATIO, dtype: float64


C:\Users\Dell\AppData\Local\Temp\ipykernel_19220\2008250836.py:1: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df["EMPLOYMENT_AGE_RATIO"] = (


In [7]:
engineered_features = [
    "LOAN_INCOME_RATIO",
    "ANNUITY_INCOME_RATIO",
    "CREDIT_GOODS_RATIO",
    "AGE_YEARS",
    "DAYS_EMPLOYED_CLEAN",
    "EMPLOYMENT_YEARS",
    "INCOME_PER_CHILD",
    "INCOME_PER_FAMILY_MEMBER",
    "CREDIT_MINUS_INCOME",
    "ANNUITY_MINUS_INCOME",
    "EMPLOYMENT_AGE_RATIO",
]

print("Engineered features:", len(engineered_features))
print()

for feature in engineered_features:
    print(
        f"{feature}: "
        f"dtype={df[feature].dtype}, "
        f"missing={df[feature].isna().sum()}"
    )

Engineered features: 11

LOAN_INCOME_RATIO: dtype=float64, missing=0
ANNUITY_INCOME_RATIO: dtype=float64, missing=12
CREDIT_GOODS_RATIO: dtype=float64, missing=278
AGE_YEARS: dtype=float64, missing=0
DAYS_EMPLOYED_CLEAN: dtype=float64, missing=55374
EMPLOYMENT_YEARS: dtype=float64, missing=55374
INCOME_PER_CHILD: dtype=float64, missing=0
INCOME_PER_FAMILY_MEMBER: dtype=float64, missing=2
CREDIT_MINUS_INCOME: dtype=float64, missing=0
ANNUITY_MINUS_INCOME: dtype=float64, missing=12
EMPLOYMENT_AGE_RATIO: dtype=float64, missing=55374


In [8]:
processed_path = (
    PROJECT_ROOT
    / "data"
    / "processed"
    / "application_train_features.csv"
)

df.to_csv(processed_path, index=False)

print("Saved:", processed_path)
print("Shape:", df.shape)

Saved: d:\credit-risk-intelligence1\data\processed\application_train_features.csv
Shape: (307511, 133)


# Day 5 — Feature Engineering Summary

### Objective

The goal of Day 5 was to create meaningful financial, demographic, household, and employment-related features from the Home Credit application dataset.

### Engineered Features

1. **LOAN_INCOME_RATIO**
   - Measures the size of the requested credit relative to applicant income.
   - Higher values indicate a larger credit burden relative to income.

2. **ANNUITY_INCOME_RATIO**
   - Measures annual loan annuity relative to total income.
   - Helps capture repayment burden.

3. **CREDIT_GOODS_RATIO**
   - Compares requested credit amount with the price of the goods being financed.
   - Can help identify differences between financed amount and goods value.

4. **AGE_YEARS**
   - Converts `DAYS_BIRTH` into approximate age in years.
   - Provides a more interpretable representation of applicant age.

5. **DAYS_EMPLOYED_CLEAN**
   - Cleans the special `365243` value in `DAYS_EMPLOYED`.
   - The placeholder is replaced with `NaN` because it does not represent a valid employment duration.

6. **EMPLOYMENT_YEARS**
   - Converts cleaned employment duration from days into years.
   - Provides a more interpretable measure of employment experience.

7. **INCOME_PER_CHILD**
   - Calculates applicant income relative to the number of children.
   - Uses `CNT_CHILDREN + 1` to avoid division by zero.

8. **INCOME_PER_FAMILY_MEMBER**
   - Calculates income relative to the applicant's family size.
   - Uses `CNT_FAM_MEMBERS + 1` to avoid division by zero.

9. **CREDIT_MINUS_INCOME**
   - Difference between credit amount and total income.
   - Captures the absolute gap between borrowing amount and income.

10. **ANNUITY_MINUS_INCOME**
    - Difference between annuity amount and total income.
    - Provides another representation of payment burden.

11. **EMPLOYMENT_AGE_RATIO**
    - Measures employment duration relative to applicant age.
    - Provides an indication of how much of the applicant's life has been spent in employment.

### Feature Engineering Results

- Original columns: **122**
- Engineered features: **11**
- Final columns: **133**
- Training rows: **307,511**

### Missing Values

Some engineered features contain missing values:

- `ANNUITY_INCOME_RATIO`: 12 missing
- `CREDIT_GOODS_RATIO`: 278 missing
- `INCOME_PER_FAMILY_MEMBER`: 2 missing
- Employment-related features: 55,374 missing

The employment-related missing values are expected because the original dataset contains a special placeholder value in `DAYS_EMPLOYED`. These missing values will be handled later during the model preprocessing stage.

### Output Dataset

The engineered training dataset was saved to:

`data/processed/application_train_features.csv`

### Overall Conclusion

Day 5 successfully transformed the raw application data into a richer feature set containing financial burden, age, employment, household, and income-related features. These engineered variables will be used in the upcoming modeling stages to improve the model's ability to identify applicants at higher risk of loan default.

In [9]:
print("DAY 5 FEATURE ENGINEERING COMPLETE")
print("Final shape:", df.shape)
print("Engineered features:", len(engineered_features))
print("Output file:", processed_path)

DAY 5 FEATURE ENGINEERING COMPLETE
Final shape: (307511, 133)
Engineered features: 11
Output file: d:\credit-risk-intelligence1\data\processed\application_train_features.csv
